packages/jsonata-paths-extractor/src/index.ts

```diff
@@ -0,0 +1,289 @@
+import * as jsonata from 'jsonata';
+interface PathContext {
+  pathNode: jsonata.ExprNode;
+  pathVariablesInScope: Record<string, PathContext>;
+  pathId: Symbol;
+  parentPathId?: Symbol;
+}
+const RootContextId = Symbol('ROOT');
+
+const UnknownContextId = Symbol('UNKNOWN');
+
+const createUnknownContext = (
+  ast: jsonata.ExprNode,
+  pathVariablesInScope: Record<string, PathContext>
+) => {
+  return {
+    pathNode: ast,
+    pathVariablesInScope: { ...pathVariablesInScope },
+    pathId: UnknownContextId,
+  };
+};
+
+const createRootContext = (ast: jsonata.ExprNode) => {
+  return {
+    pathNode: ast,
+    pathVariablesInScope: {},
+    pathId: RootContextId,
+  };
+};
+
+const reduceToPathString = (
+  pathContext: PathContext,
+  stopAtPosition?: number
+): string => {
+  return pathContext.pathNode
+    .steps!.filter(step =>
+      typeof stopAtPosition === 'number'
+        ? step.position! < stopAtPosition
+        : true
+    )
+    .reduce((pathString, step, index, array) => {
+      if (index === 0 && step.type === 'variable') {
+        if (step.value === '$') {
+          // $$. means reset context to root
+          pathContext.parentPathId = RootContextId;
+          return '';
+        } else if (step.value === '') {
+          return '';
+        } else if (pathContext.pathVariablesInScope[step.value]) {
+          //look for variable in scope
+          return (
+            pathString +
+            reduceToPathString(pathContext.pathVariablesInScope[step.value])
+          );
+        } else {
+          // end early for unknown variables
+          array.splice(index + 1);
+          return pathString;
```

Comment:
Is splicing an array in the middle of reducing it a safe operation? Normally you don't want to modify things you are iterating.

Reply:
It does modify the original array, so it is somewhat unsafe, but in this case I think it is okay to modify the current array since the steps that are being removed are not needed. I could make a clone of steps before iterating if I needed to though. Specifically during the reduce, this is the only way I know how to break out of the reduce early. The behaviour is defined in the last bullet point here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#behavior_during_array_mutations